File: app/src/main/java/cn/nekocode/baseframework/BaseActivity.java
package cn.nekocode.baseframework;

import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.support.v7.app.AppCompatActivity;
import android.view.MenuItem;

import java.lang.ref.WeakReference;
import java.util.ArrayList;

public class BaseActivity extends AppCompatActivity {
    private static ArrayList<Handler> handlers = new ArrayList<Handler>();
    protected MyHandler handler = new MyHandler(this);

    public BaseActivity _this;

    public static void addHandler(Handler handler) {
        handlers.add(handler);
    }

    public static void deleteHandler(Handler handler) {
        handlers.remove(handler);
    }

    public static void removeAll() {
        handlers.clear();
    }

    public static void broadcast(Message message) {
        for (Handler handler : handlers) {
            Message msg = new Message();
            msg.copyFrom(message);
            handler.sendMessage(msg);
        }
    }

    public void sendMsg(Message message) {
        Message msg = new Message();
        msg.copyFrom(message);
        handler.sendMessage(msg);
    }

    public void sendMsgDelayed(Message message, int delayMillis) {
        Message msg = new Message();
        msg.copyFrom(message);
        handler.sendMessageDelayed(msg, delayMillis);
    }

    public void runDelayed(Runnable runnable, int delayMillis) {
        Message msg = new Message();
        msg.what = -101;
        msg.arg1 = -102;
        msg.arg2 = -103;
        msg.obj = runnable;
        handler.sendMessageDelayed(msg, delayMillis);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        _this = this;
        addHandler(handler);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch(item.getItemId()) {
            case android.R.id.home:
                this.finish();
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onResume() {
        super.onResume();
    }

    @Override
    public void onPause() {
        super.onPause();
    }

    @Override
    public void finish() {
        deleteHandler(handler);
        super.finish();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
    }

    static class MyHandler extends Handler {
        private WeakReference<BaseActivity> mOuter;

        public MyHandler(BaseActivity activity) {
            mOuter = new WeakReference<BaseActivity>(activity);
        }

        @Override
        public void handleMessage(Message msg) {
            final BaseActivity outer = mOuter.get();
            if (outer == null) {
                return;
            }

            if(msg.what==-101 && msg.arg1==-102 && msg.arg2==-103) {
                ((Runnable) msg.obj).run();
                return;
            }

            outer.handler(msg);
        }
    }

    public void handler(Message msg) {

    }
}

File: app/src/main/java/cn/nekocode/baseframework/ui/activity/helper/BaseActivity.kt
package cn.nekocode.baseframework.ui.activity.helper

import android.app.Activity
import android.os.Bundle
import android.os.Handler
import android.os.Message
import android.support.v4.app.ActionBarDrawerToggle
import android.support.v7.app.AppCompatActivity
import android.view.MenuItem

import java.lang.ref.WeakReference
import java.util.ArrayList
import kotlin.properties.Delegates

open abstract class BaseActivity : AppCompatActivity() {
    companion object {
        private val handlers = ArrayList<MyHandler>()

        fun addHandler(handler: MyHandler) {
            handlers.add(handler)
        }

        fun deleteHandler(handler: MyHandler) {
            handlers.remove(handler)
        }

        fun removeAll() {
            handlers.clear()
        }

        fun broadcast(message: Message) {
            for (handler in handlers) {
                val msg = Message()
                msg.copyFrom(message)
                handler.sendMessage(msg)
            }
        }
    }

    protected val handler: MyHandler by Delegates.lazy {
        MyHandler(this)
    }

    public val _this: BaseActivity by Delegates.lazy {
        this!!
    }

    public fun sendMsg(message: Message) {
        val msg = Message()
        msg.copyFrom(message)
        handler.sendMessage(msg)
    }

    public fun sendMsgDelayed(message: Message, delayMillis: Int) {
        val msg = Message()
        msg.copyFrom(message)
        handler.sendMessageDelayed(msg, delayMillis.toLong())
    }

    public fun runDelayed(runnable: ()->Unit, delayMillis: Int) {
        val msg = Message()
        msg.what = -101
        msg.arg1 = -102
        msg.arg2 = -103
        msg.obj = runnable
        handler.sendMessageDelayed(msg, delayMillis.toLong())
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        addHandler(handler)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.getItemId()) {
            android.R.id.home -> this.finish()
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onResume() {
        super.onResume()
    }

    override fun onPause() {
        super.onPause()
    }

    override fun finish() {
        deleteHandler(handler)
        super.finish()
    }

    override fun onDestroy() {
        super.onDestroy()
    }

    inner class MyHandler : Handler {
        private val mOuter: WeakReference<BaseActivity>

        constructor(activity: BaseActivity) {
            mOuter = WeakReference(activity)
        }

        override fun handleMessage(msg: Message) {
            val outer = mOuter.get() ?: return

            if (msg.what == -101 && msg.arg1 == -102 && msg.arg2 == -103) {
                (msg.obj as (()->Unit)).invoke()
                return
            }

            outer.handler(msg)
        }
    }

    public abstract fun handler(msg: Message)
}
